src/charm.py

```diff
@@ -5,67 +5,139 @@
+        return self.model.get_relation(PEER)
 
     def _on_install(self, _) -> None:
-        """Handler for on_install event."""
-        self.unit.status = self.snap.install_kafka_snap()
+        """Handler for `install` event."""
+        if self.snap.install():
+            self.kafka_config.set_kafka_opts()
+            self.unit.status = WaitingStatus("waiting for zookeeper relation")
+        else:
+            self.unit.status = BlockedStatus("unable to install kafka snap")
 
     def _on_leader_elected(self, _) -> None:
-        return
-
-    def _on_cluster_relation_joined(self, _) -> None:
-        return
-
-    def _on_config_changed(self, _) -> None:
-        """Handler for config_changed event."""
-        self._start_service()
-
-    def _start_service(self) -> None:
-        """Starts a service made available in the snap - `kafka` or `zookeeper`."""
-        self.unit.status = self.snap.start_snap_service(snap_service="kafka")
-
-    def _on_get_properties_action(self, event) -> None:
-        """Handler for users to copy currently active config for passing to `juju config`."""
-        msg = self.snap.get_merged_properties(property_label="server")
-        event.set_results({"properties": msg})
-
-    def _on_get_snap_apps_action(self, event) -> None:
-        """Handler for users to retrieve the list of available Kafka snap commands."""
-        msg = self.snap.get_kafka_apps()
-        event.set_results({"apps": msg})
+        """Handler for `leader_elected` event, ensuring sync_passwords gets set."""
+        sync_password = self.kafka_config.sync_password
+        if not sync_password:
+            self.peer_relation.data[self.app].update(
+                {
+                    "sync_password": "".join(
+                        [secrets.choice(string.ascii_letters + string.digits) for _ in range(32)]
+                    )
+                }
+            )
+
+    def _on_zookeeper_joined(self, event: RelationJoinedEvent) -> None:
+        """Handler for `zookeeper_relation_joined` event, ensuring chroot gets set."""
+        if self.unit.is_leader():
+            event.relation.data[self.app].update({"chroot": "/" + self.app.name})
+
+    def _on_zookeeper_created(self, event: EventBase) -> None:
+        """Handler for `zookeeper_relation_created` event."""
+        # if missing zookeeper_config, required data might not be set yet
+        if not zookeeper_connected(charm=self):
+            event.defer()
+            return
+
+        # for every new ZK relation, start kafka service
+        self._on_start(event=event)
+
+    def _on_zookeeper_broken(self, _: RelationEvent) -> None:
+        """Handler for `zookeeper_relation_departed/broken` events."""
+        # if missing zookeeper_config, there is no required ZooKeeper relation, block
+        if not zookeeper_connected(charm=self):
+            logger.info("stopping snap service")
+            self.snap.stop_snap_service(snap_service=CHARM_KEY)
```

Comment:
Should it be a restart snap service to not have the service stopped if zookeper is connected again?

Reply:
So I think the logic here, was to ensure a `start_snap_service` for every `zookeeper_created` event, and a `stop_snap_service` for every `zookeeper_departed/broken` event where relation data disappears. 
Running `stop_snap_service` repeatedly doesn't start it, it will do nothing.

Given that, I'm not sure what you mean here, are you suggesting that on `relation_broken` events I should restart the service rather than stop?

Reply:
Ah, sorry Marc. I totally missed L92. Thanks for the explanation. It answers my question.

Reply:
All gud!